src/java.base/share/native/libjava/jni_util.c

```diff
@@ -733,31 +730,38 @@ InitializeEncoding(JNIEnv *env, const char *encname)
             strcmp(encname, "utf-16le") == 0) {
             fastEncoding = FAST_CP1252;
         } else {
-            jboolean exe;
-            jstring enc = (*env)->NewStringUTF(env, encname);
-            if (enc == NULL)
+            charsetname = encname;
+            fastEncoding = NO_FAST_ENCODING;
+        }
+        while (charsetname != NULL) {
+            jstring enc = (*env)->NewStringUTF(env, charsetname);
+            if (enc == NULL) {
+                fastEncoding = NO_ENCODING_YET;
                 return;
+            }
+            jboolean exc;
+            jvalue charset = JNU_CallStaticMethodByName(
+                    env, &exc,
+                    "java/nio/charset/Charset",
+                    "forName",
+                    "(Ljava/lang/String;)Ljava/nio/charset/Charset;",
+                    enc);
+            if (exc) {
+                (*env)->ExceptionClear(env);
+            }
+            (*env)->DeleteLocalRef(env, enc);
 
-            if ((jboolean) JNU_CallStaticMethodByName (
-                                            env, &exe,
-                                            "java/nio/charset/Charset",
-                                            "isSupported",
-                                            "(Ljava/lang/String;)Z",
-                                            enc).z == JNI_TRUE) {
-                fastEncoding = NO_FAST_ENCODING;
-                jnuEncoding = (jstring)(*env)->NewGlobalRef(env, enc);
-            } else {
-                // jnuEncoding falls back to UTF-8
-                jstring utf8 = (*env)->NewStringUTF(env, "UTF-8");
-                if (utf8 == NULL) {
-                    (*env)->DeleteLocalRef(env, enc);
-                    return;
-                }
+            if (!exc && charset.l != NULL) {
+                jnuEncoding = (*env)->NewGlobalRef(env, charset.l);
+                (*env)->DeleteLocalRef(env, charset.l);
+                break;
```

Comment:
Could return immediately

Reply:
No, then the method id cache would not get initialized below. But I will add a comment.

Reply:
You are right, those init code should follow.